3rdparty/openvx/include/ivx.hpp

```diff
@@ -0,0 +1,1529 @@
+    {}
+
+    explicit RefWrapper(T r, bool retainRef = false) : ref(0), refcount(0)
+    { reset(r, retainRef); }
+
+    RefWrapper(const RefWrapper& r) : ref(r.ref), refcount(r.refcount)
+    { addRef(); }
+
+    operator T() const
+    { return ref; }
+
+    operator vx_reference() const
+    { return castToReference(ref); }
+
+    void reset(T r, bool retainRef = false)
+    {
+        release();
+        ref = r;
+#ifdef VX_VERSION_1_1
+        if(retainRef) addRef();
+#else
+        // if 'retainRef' -just don't use ref-counting for v 1.0
+        if(!retainRef) refcount = new int(1);
+#endif
+        checkRef();
+    }
+
+    void reset()
+    { release(); }
+
+    T detach()
+    {
+        T tmp = ref;
+        ref = 0;
+        release();
+        return tmp;
+    }
+
+    RefWrapper& operator=(RefWrapper r)
+    {
+        using std::swap;
+        swap(ref, r.ref);
+        swap(refcount, r.refcount);
+        return *this;
+    }
+
+    bool operator !() const
+    { return ref == 0; }
+
+#ifdef IVX_USE_CXX98
+    template<typename C>
+    C get() const
+    {
+        typedef int static_assert_context[is_same<C, Context>::value ? 1 : -1];
+        vx_context c = vxGetContext(castToReference(ref));
+        // vxGetContext doesn't increment ref count, let do it in wrapper c-tor
+        return C(c, true);
+    }
+#else
+    template<typename C = Context, typename = typename std::enable_if<std::is_same<C, Context>::value>::type>
+    C getContext() const
+    {
+        vx_context c = vxGetContext(castToReference(ref));
+        // vxGetContext doesn't increment ref count, let do it in wrapper c-tor
+        return C(c, true);
+    }
+#endif // IVX_USE_CXX98
+
+protected:
+    T ref;
+    int* refcount;
+
+    void addRef()
+    {
+#ifdef VX_VERSION_1_1
+        if(ref) IVX_CHECK_STATUS(vxRetainReference(castToReference(ref)));
+#else //TODO: make thread-safe
+        if(refcount) ++(*refcount);
+#endif
+    }
+
+    void release()
+    {
+#ifdef VX_VERSION_1_1
+        if(ref) RefTypeTraits<T>::release(ref);
+#else //TODO: make thread-safe
+        if(refcount && --(*refcount) == 0)
+        {
+            if(ref) RefTypeTraits<T>::release(ref);
+            ref = 0;
+            delete refcount;
+            refcount = 0;
+        }
+#endif
+    }
+
+    void checkRef() const
+    {
+        IVX_CHECK_REF(ref);
+        vx_enum type;
+        IVX_CHECK_STATUS(vxQueryReference((vx_reference)ref, VX_REF_ATTRIBUTE_TYPE, &type, sizeof(type)));
+        if (type != vxTypeEnum) throw WrapperError("incompatible reference type");
+    }
+
+    ~RefWrapper()
+    { release(); }
+};
+
+#define IVX_REF_STD_CTORS_AND_ASSIGNMENT(Class) \
+    Class() : RefWrapper() {} \
+    explicit Class(Class::vxType _ref, bool retainRef = false) : RefWrapper(_ref, retainRef) {} \
+    Class(const Class& _obj) : RefWrapper(_obj) {} \
+    \
+    Class& operator=(Class _obj) { using std::swap; swap(ref, _obj.ref); return *this; }
```

Comment:
actually this is ["unified assignment operator"](http://en.cppreference.com/w/cpp/language/operators#Assignment_operator) that covers both  *copy assignment* and *move assignment*.